numpy/core/src/common/array_assign.c

```diff
@@ -84,14 +84,27 @@ broadcast_error: {
 
 /* See array_assign.h for parameter documentation */
 NPY_NO_EXPORT int
-raw_array_is_aligned(int ndim, char *data, npy_intp *strides, int alignment)
+raw_array_is_aligned(int ndim, npy_intp *shape,
+                     char *data, npy_intp *strides, int alignment)
 {
     if (alignment > 1) {
         npy_intp align_check = (npy_intp)data;
-        int idim;
-
-        for (idim = 0; idim < ndim; ++idim) {
-            align_check |= strides[idim];
+        int i;
+
+        for (i = 0; i < ndim; i++) {
+#if NPY_RELAXED_STRIDES_CHECKING
+            /* skip dim == 1 as it is not required to have stride 0 */
+            if (shape[i] > 1) {
+                /* if shape[i] == 1, the stride is never used */
+                align_check |= strides[i];
```

Comment:
I think there is an assumption of multiples of powers of two, together with twos complement arithmetic, built in here. AFAICT, there are no false positive, but perhaps some false negatives. @juliantaylor @seberg Thoughts?

Reply:
I don't see any arithmetic here, only bitwise OR operations. Whether the integer is considered signed or unsigned doesn't come into play.

Reply:
Well the bitwise or is used together with arrhythmic stuff later since it is used in `npy_is_aligned`. Did not think about it long, but I think you are right Chuck. I guess `npy_is_aligned` might explain it a bit more?

Reply:
Not sure if twos complement has anything to do with it though.

Reply:
"arrythmic"? :-)

More seriously, I don't know what you are bothering about. First `data` is cast to `npy_intp`(this is an exact bitcast), some bitwise operations are done on the signed integer (they are agnostic wrt signedness, they just operate on individual bits), then the result is cast back to `void*` (this is an exact bitcast), then it is finally cast to `npy_uintp`.

There is zero reason to believe that the intermediate cast to `npy_intp` is any potential cause for trouble.

Reply:
Casting unsigned integers to signed integers is implementation-defined

Reply:
> Not sure if twos complement has anything to do with it though.

For example, ones complement turns even into odd when cast to unsigned.

Reply:
You mean when ORing `strides[i]` with `align_check`? I would be surprised if the compiler produced anything else than a pure bitwise OR here. You may want to enable conversion warnings to see if it says anything.

Reply:
Anyway, I don't even know why I'm commenting on this PR. I have other things to do.

Reply:
> I guess npy_is_aligned might explain it a bit more?

That function is pretty straight forward, but I think what happens with the or's is that we are guaranteed the largest power of two that divides all the possible addresses. I think what we would really like is the GCD of all the used addresses, which reduces to the GCD of the address and the stride, and then check if that is divisible by the alignment. But that is more complicated.

Reply:
> Again, these are pure bitwise casts here.

Yes, I know. But one does lose some information with two complement. Note, for instance, on a three bit machine, 3 (011) and -3 (101) are both divisible by 3 but 111 is not.

Reply:
Well, if this errs too much on the safe side (no idea if it does). I guess placing  the `npy_is_aligned` check into the loop here instead should work and probably not be a lot of overhead. (Plus this is the flags calculation, right? so it doesn't happen very often anyway). Or am I missing something? Plus it might be a bit less mind boggling.

Reply:
I was planning to sleep on it :) I think when all is done, it will be OK for its intended application, but I want to understand it...

Reply:
I think you're right that this assumes a twos-complement representation of signed integers, even though the C standard explicitly allows others like ones-complement. I'm not sure we care, because in practice all modern computers use two's complement, and because all this pointer bit-twiddling is C-implementation-dependent anyway.

Here's my understanding: We are trying to compute whether `(data + n*stride)%alignment == 0` for all integers `n`. First, assuming alignment is a power of two, and assuming twos-complement representation, simplify the `%alignment == 0` to checking if the lowest log2(alignment) bits are 0. Next, use the fact that in twos complement representation, if `x` low order bits of `data` and `stride` are 0, then the `x` lower bits of `data + n*stride` must also be 0 for all `n` because of how binary addition works. Neither simplification is valid for negative values in ones-complement representation.

Casting the intp `stride` to uintp` and using uintp would avoid the signed twos-complement assumption, and the algorithm still works.

However, the representation/cast of `data` from pointer to uint is implementation defined too, so we are violating the standard anyway anytime we use bitwise ops on pointer values. (only arithmetic is allowed, I think). I don't see any alternative though, as far as I see there is no totally standard-compliant way to test alignment.